### PR TITLE
Added isVisible function to fix Travis CI. fixes #934

### LINKED ISSUE #934: Fix error on travis CI so the test suite can be executed
## Description
As a contributor,
I need to have my changes tested on travis CI,
so that I can assure that my changes dont break anything in current version of code.

## Mocks
image

### Update [Required]
- [ ] .travis.yml file to fix the build so the test suite can run properly

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
1 hour

### PR DESCRIPTION
**Pull request by Harirai**
# Description
Configured .travis.yml to fix `xvfb` and run organization app only, for now.
Added `isVisible` function to avoid NoSuchElementExcepltion.

Fixes #934 

# Type of Change:

- Code
- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
This has been tested by making relevant changes locally and triggering Travis build for the changes by pushing the commits on my fork.

# Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [ ] Any dependent changes have been merged 

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings 
- [ ] My PR currently breaks something (Later Commits yet to be added, after approval of mentor)

### CODE REVIEW

**Comment by Harirai** [comment]
@Monal5031 I have made the initial commits (Added the function and configured .travis.yml).
I don't know why an extra commit is appearing. (It has been already merged though, and I rebased my fork accordingly, but still, it is appearing in the PR :sweat_smile: )

On your approval, I'll add rest of the commits to fix corresponding apps.

**Comment by snitin315** [comment]
> I don't know why an extra commit is appearing. (It has been already merged though, and I rebased my fork accordingly, but still, it is appearing in the PR  )
 
Keeping a clean git history is difficult actually, to avoid such behavior you can always reset your branch with upstream before starting work on a new issue.

`git reset --hard upstream/develop`

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # Contributing Guidelines
-* You can join our [slack channel](http://systers.io/slack-systers-opensource/).  Each active repo has its own channel to direct questions to (for example #powerup or #portal).  
+* You can join our [ AnitaB.org Open Source Zulip Channel](https://anitab-org.zulipchat.com/).  Each active repo has its own channel to direct questions to (for example [#VMS](https://anitab-org.zulipchat.com/#narrow/stream/222539-vms), [#powerup](https://anitab-org.zulipchat.com/#narrow/stream/222537-powerup) or [#portal](https://anitab-org.zulipchat.com/#narrow/stream/222540-portal)).
```
Comment (snitin315) [inline CONTRIBUTING.md:2]: Please remove unnecessary changes from this file.

Comment (Harirai) [reply CONTRIBUTING.md:2]: @snitin315 As I said these changes have been merged. These were present in [#964](https://github.com/anitab-org/vms/pull/964).
I don't know how to "undo" it :sweat_smile:

File: vms/organization/tests/test_services.py [hunk 2]
```diff
@@ -110,12 +113,9 @@ def setup_test_data(cls):
 
         cls.o1.save()
         cls.o2.save()
-        city_name = 'Bothell'
-        state_name = 'Washington'
-        country_name = 'United States'
-        country = get_country_by_name(country_name)
-        state = get_state_by_name(state_name)
-        city = get_city_by_name(city_name)
+        country = create_second_country()
```
Comment (Monal5031) [inline vms/organization/tests/test_services.py:116]: Why did you change this?

File: vms/pom/pages/basePage.py [hunk 3]
```diff
@@ -14,14 +20,27 @@ class BasePage(object):
     def __init__(self, driver):
         self.driver = driver
 
+    # New function to check if Element is visible before calling it
+    def is_visible(self, element, timeout=5):
+        try:
+            wait(self.driver,
```
Comment (Monal5031) [inline vms/pom/pages/basePage.py:26]:
```py
wait(self.driver, timeout)\
	.until(
  		ec.visibility_of_element_located(
        	(By.XPATH, element)
        )
	)
```

File: vms/pom/pages/basePage.py [hunk 4]
```diff
@@ -14,14 +20,27 @@ class BasePage(object):
...
+                 timeout).until(ec.visibility_of_element_located((By.XPATH,
+                                                                  element)))
+            return True
+        except TimeoutException:
+            return NoSuchElementException
```
Comment (Monal5031) [inline vms/pom/pages/basePage.py:31]: Just return `False` your use of this function already raises this error on it's own

File: vms/volunteer/tests/test_services.py [hunk 5]
```diff
@@ -18,12 +20,12 @@
 class VolunteerMethodTests(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        country_name = 'India'
-        country = get_country_by_name(country_name)
-        state_name = 'Uttarakhand'
-        state = get_state_by_name(state_name)
-        city_name = 'Roorkee'
-        city = get_city_by_name(city_name)
+        # country_name = 'India'
```
Comment (Monal5031) [inline vms/volunteer/tests/test_services.py:23]: Why change?

File: vms/volunteer/tests/test_services.py [hunk 6]
```diff
@@ -232,12 +234,12 @@ def test_search_volunteers(self):
 class DeleteVolunteerTest(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        country_name = 'United States'
-        country = get_country_by_name(country_name)
-        state_name = 'Washington'
-        state = get_state_by_name(state_name)
-        city_name = 'Bothell'
-        city = get_city_by_name(city_name)
+        # country_name = 'United States'
```
Comment (Monal5031) [inline vms/volunteer/tests/test_services.py:237]: Why this change?

File: CONTRIBUTING.md (hunk 1 above)
Comment (Kishan-Dhakan) [reply CONTRIBUTING.md:2]: This may have happened if you branched to IsVisibleFunction directly from "Harirai-Contributing". It wouldn't happen if you had branched from your develop (unless you already merged Harirai-Contributing to your develop)

Comment (yashshah2820) [reply CONTRIBUTING.md:2]: try rebasing the commits

File: vms/pom/pages/basePage.py (hunk 4 above)
Comment (Harirai) [reply vms/pom/pages/basePage.py:31]: The `TimeoutException` is raised when the element is not found even after 5s, if that happens then, it must be because Element is not there.
`TimeoutException` might be misleading when there's no such element instead.

File: vms/pom/pages/basePage.py (hunk 3 above)
Comment (Harirai) [reply vms/pom/pages/basePage.py:26]: Thanks :) I'll do it this way.

File: vms/organization/tests/test_services.py (hunk 2 above)
Comment (Harirai) [reply vms/organization/tests/test_services.py:116]: There was no country in the `test_vms` database when I was running the test on this app only.
So to create an entry in the database.

File: vms/pom/pages/basePage.py (hunk 4 above)
Comment (Monal5031) [reply vms/pom/pages/basePage.py:31]: You are returning `NoSuchElementException` and not raising it, so returning False instead will also do the same thing as per your uses of this function in other places

File: vms/volunteer/tests/test_services.py (hunk 5 above)
Comment (Harirai) [reply vms/volunteer/tests/test_services.py:23]: Same reason for all the `create_` functions used, these were not present in the database.
And due to This Not Found error was thrown.

File: vms/pom/pages/basePage.py (hunk 4 above)
Comment (Harirai) [reply vms/pom/pages/basePage.py:31]: Okay sir got it. Sorry for this silly error. :sweat_smile:

File: vms/organization/tests/test_services.py (hunk 2 above)
Comment (Monal5031) [reply vms/organization/tests/test_services.py:116]: Did you check it locally or CI?

File: vms/pom/pages/basePage.py (hunk 4 above)
Comment (Monal5031) [reply vms/pom/pages/basePage.py:31]: Only Monal is fine.

File: vms/volunteer/tests/test_services.py (hunk 5 above)
Comment (Monal5031) [reply vms/volunteer/tests/test_services.py:23]: Check it on CI first, locally it might be due to the download of cities_light

File: vms/organization/tests/test_services.py (hunk 2 above)
Comment (Harirai) [reply vms/organization/tests/test_services.py:116]: I checked it both ways.
Actually the issue is when we check apps individually, I needed to use `create_` whenever not found exception was there.
But if we run all the apps together, they have common `test_vms` database. So we need to create it only once in any app. And we'll be able to call it in other apps too.

Comment (Monal5031) [reply vms/organization/tests/test_services.py:116]: Makes sense, then create new Country etc in setUp specifically for these tests and use them only here and in no other app

File: vms/volunteer/tests/test_services.py (hunk 5 above)
Comment (Harirai) [reply vms/volunteer/tests/test_services.py:23]: Yes, I checked on CI.
I think `cities_light` is not used in tests and does not affect tests.
That's why `create_` functions were previously created. (I didn't write `create_` functions I only called them.)

File: vms/organization/tests/test_services.py (hunk 2 above)
Comment (Harirai) [reply vms/organization/tests/test_services.py:116]: @Monal5031  Sorry I didn't get you exactly.

The latest build for this PR is failing because I called `create_` in volunteer but as This build runs two apps, country etc. were created in prev (organisation app).
How should I manage this? Should I call `get_by` before creating country etc. ?

**Comment by yashshah2820** [comment]
What's the update on this issue, could @Harirai any gif or image related to the fix

**Comment by Harirai** [comment]
@yashshah2820 @satya7289 I'm not able to fix this. If someone can help me out regarding the changes, please do so.
Otherwise, I think it's better to assign this issue to someone else because this needs to be fixed quickly.

**Comment by satya7289** [comment]
Ok

**Comment by Harirai** [comment]
I am closing this PR because Travis is not fixed by this, and this is causing hindrance in its progress.
The issue mentioned needs to be divided into two issues.
One which fixes Travis and next which fixes failing tests.
